### ISSUE TITLE
Can no longer derive Union for union types with lifetimes.

### ISSUE THREAD

**Issue by Snnappie** [async-graphql/async-graphql#311]
I am attempting to upgrade a project which was pinned to version `1.17.2` to version 2. After changing various `#[SimpleObject]` to `#[derive(SimpleObject)]`, things were looking good until I came across this issue. Consider the following code:

```rust

// This fails to compile, complaining of undeclared lifetime name `'b`.
#[derive(async_graphql::Union)]
enum UnionLt<'b> {
    A(A<'b>),
    B(B<'b>),
}

struct A<'a> {
    field: &'a str,
}

struct B<'a> {
    data: &'a str,
}

// Without the derive(Union), these two compile fine.
#[async_graphql::Object]
impl<'a> A<'a> {
    async fn field(&self) -> &str {
        self.field
    }
}

#[async_graphql::Object]
impl<'a> B<'a> {
    async fn data(&self) -> &str {
        self.data
    }
}
```
This was compiling (and working fine) on 1.17.2.

**Comment by Snnappie** [comment]
So I dug into this some more with `cargo expand`. The relevant snippet of code is here:
```rust
    enum UnionLt<'b> {
        A(A<'b>),
        B(B<'b>),
    }
    const _: fn() = || {
        trait AmbiguousIfMoreThanOne<A> {
            fn some_item() {}
        }
        {
            #[allow(dead_code)]
            struct Token;
            impl<T: ?Sized + async_graphql::ObjectType> AmbiguousIfMoreThanOne<Token> for T {}
        }
/////////////// VVVVV Error is here -- replace this 'b with '_ and it compiles
        let _ = <A<'b> as AmbiguousIfMoreThanOne<_>>::some_item;
    };
    // ... snip
    const _: fn() = || {
        trait AmbiguousIfMoreThanOne<A> {
            fn some_item() {}
        }
        {
            #[allow(dead_code)]
            struct Token;
            impl<T: ?Sized + async_graphql::ObjectType> AmbiguousIfMoreThanOne<Token> for T {}
        }
//////////////////VVVV here as well
        let _ = <B<'b> as AmbiguousIfMoreThanOne<_>>::some_item;
    };
```

I can replace the `#[derive(Union)]` with the expanded code and replace the marked `'b` lifetimes with `'_` and now it compiles fine.

**Comment by Snnappie** [comment]
Some more digging - it is due to this section of code: https://github.com/async-graphql/async-graphql/blob/bc5cf2f2a2b5f0cb71e392a07a21332a8c23b1bc/derive/src/union.rs#L71-L93

In particular the static assertions on lines 73 and 84. If I remove them, my code compiles. I'm not sure if they are necessary. I spent some time trying to trick the generated code to have the appropriate lifetime specified, but I was unsuccessful. Seems like you could replace any lifetime references in the type identifier with `'_`, but I'm not familiar enough with `syn`, etc. to implement that.

**Comment by sunli829** [comment]
@Snnappie Thank you for helping me find the cause of the problem, I fixed it in `v2.0.3`. 😁

**Comment by Snnappie** [comment]
@sunli829 Perfect, my code compiles with `2.0.3` now. Thanks!
